fdbserver/include/fdbserver/IEncryptionKeyProvider.actor.h

```diff
@@ -0,0 +1,281 @@
+		s.secret = StringRef(s.arena(), &secret, 1);
+		return s;
+	}
+
+	Future<EncryptionKey> getByRange(const KeyRef& begin, const KeyRef& end) override {
+		EncryptionKeyRef k;
+		k.id = end.empty() ? 0 : *(end.end() - 1);
+		return getSecrets(k);
+	}
+
+	uint8_t xorWith;
+};
+
+// Key provider to provider cipher keys randomly from a pre-generated pool. Use for testing.
+class RandomEncryptionKeyProvider : public IEncryptionKeyProvider {
+public:
+	RandomEncryptionKeyProvider() {
+		for (unsigned i = 0; i < NUM_CIPHER; i++) {
+			BlobCipherDetails cipherDetails;
+			cipherDetails.encryptDomainId = i;
+			cipherDetails.baseCipherId = deterministicRandom()->randomUInt64();
+			cipherDetails.salt = deterministicRandom()->randomUInt64();
+			cipherKeys[i] = generateCipherKey(cipherDetails);
+		}
+	}
+	virtual ~RandomEncryptionKeyProvider() = default;
+
+	bool shouldEnableEncryption() const override { return true; }
+
+	Future<EncryptionKey> getSecrets(const EncryptionKeyRef& key) override {
+		ASSERT(key.cipherHeader.present());
+		EncryptionKey s = key;
+		s.cipherKeys.cipherTextKey = cipherKeys[key.cipherHeader.get().cipherTextDetails.encryptDomainId];
+		s.cipherKeys.cipherHeaderKey = cipherKeys[key.cipherHeader.get().cipherHeaderDetails.encryptDomainId];
+		return s;
+	}
+
+	Future<EncryptionKey> getByRange(const KeyRef& /*begin*/, const KeyRef& /*end*/) override {
+		EncryptionKey s;
+		s.cipherKeys.cipherTextKey = getRandomCipherKey();
+		s.cipherKeys.cipherHeaderKey = getRandomCipherKey();
+		return s;
+	}
+
+private:
+	Reference<BlobCipherKey> generateCipherKey(const BlobCipherDetails& cipherDetails) {
+		static unsigned char SHA_KEY[] = "3ab9570b44b8315fdb261da6b1b6c13b";
+		Arena arena;
+		StringRef digest = computeAuthToken(reinterpret_cast<const unsigned char*>(&cipherDetails.baseCipherId),
+		                                    sizeof(EncryptCipherBaseKeyId),
+		                                    SHA_KEY,
+		                                    AES_256_KEY_LENGTH,
+		                                    arena);
+		return makeReference<BlobCipherKey>(cipherDetails.encryptDomainId,
+		                                    cipherDetails.baseCipherId,
+		                                    digest.begin(),
+		                                    AES_256_KEY_LENGTH,
+		                                    cipherDetails.salt,
+		                                    std::numeric_limits<int64_t>::max() /* refreshAt */,
+		                                    std::numeric_limits<int64_t>::max() /* expireAt */);
+	}
+
+	Reference<BlobCipherKey> getRandomCipherKey() {
+		return cipherKeys[deterministicRandom()->randomInt(0, NUM_CIPHER)];
+	}
+
+	static constexpr int NUM_CIPHER = 1000;
+	Reference<BlobCipherKey> cipherKeys[NUM_CIPHER];
+};
+
+// Key provider which extract tenant id from range key prefixes, and fetch tenant specific encryption keys from
+// EncryptKeyProxy.
+class TenantAwareEncryptionKeyProvider : public IEncryptionKeyProvider {
```

Comment:
I see value in keeping `encryption` as a feature independent from `Tenant` terminology; in current code, Tenant boundary == encryption boundary, however, the encryption codebase works on `EncryptionDomainId`, in future it allows the code to evolve to introduce `domains` which aren't bounded by Tenants; for instance: 


How about : `EncryptionDomainAwareKeyProvider` instead? Also, for uniformity lets use  `EncryptUtils::EncryptCipherDomainId` as identifier instead of TenantId if possible

Reply:
With per-tenant encryption, apparently encryption is not independent from tenant. And this EncryptionKeyProvider is one place where the two concept interoperates. If we are super into abstraction, maybe we could have a `IEncryptionDomainProvider` interface and its implementation `PerTenantEncryptionDomainProvider` which can plugin into the EncryptionKeyProvider and hides the tenant prefix knowledge, but I think that's overdoing it.

That said, I think `PerTenantEncryptionKeyProvider` is a better name. What do you think?

Reply:
With per-tenant encryption, apparently encryption is not independent from tenant. And this EncryptionKeyProvider is one place where the two concept interoperates. If we are super into abstraction, maybe we could have a `IEncryptionDomainProvider` interface and its implementation `PerTenantEncryptionDomainProvider` which can plugin into the EncryptionKeyProvider and hides the tenant prefix knowledge, but I think that's overdoing it.

That said, I think `PerTenantEncryptionKeyProvider` is a better name. What do you think?

Reply:
Sounds good, the current name makes sense. 
Thanks